File: icicle-gradle/src/main/java/net/iceyleagons/gradle/IciclePlugin.java
package net.iceyleagons.gradle;

import com.amihaiemil.eoyaml.Yaml;
import com.amihaiemil.eoyaml.YamlMappingBuilder;
import com.amihaiemil.eoyaml.YamlNode;
import lombok.SneakyThrows;
import lombok.val;
import lombok.var;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.artifacts.Configuration;
import org.gradle.api.artifacts.Dependency;
import org.gradle.api.artifacts.ExternalModuleDependency;
import org.gradle.api.file.DuplicatesStrategy;
import org.gradle.api.tasks.Copy;
import org.gradle.api.tasks.compile.JavaCompile;
import org.gradle.jvm.tasks.Jar;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;

public class IciclePlugin implements Plugin<Project> {
    public void apply(Project target) {
        // Apply java plugin for "default" tasks and extensions.
        target.getPluginManager().apply("java");

        val addonData = setupAddonData(target);

        // Dependency types: icicle, icicleShadow, shadow, runtimeDownload
        val icicleConfiguration = setupDependencies(target, addonData);
        setupShadow(target, icicleConfiguration);
        setupRuntimeDownload(target, addonData);

        // Setup dependencies: #igloo(), #iglooSnapshots(), #sonatype(), #codemc(), #paper(), #spigot()
        setupRepository(target);

        setupDependencyAliases(target);

        // Icicle.yml generation setup
        setupIcicleGeneration(target, addonData);

        // Set the project encoding to UTF-8 & Minecraft-specific: modify plugin.yml to include Icicle in it's dependencies
        setupUTF8(target);
        setupPluginYML(target);
    }

    private void setupPluginYML(Project target) {
        target.getTasks().getByName("processResources").doFirst(task -> {
            try {
                val pluginYml = target.file("src/main/resources/plugin.yml");
                if (pluginYml.exists()) {
                    val yamlBuilder = editPluginYml(pluginYml);

                    if (!pluginYml.delete() || !pluginYml.createNewFile()) return;
                    try (FileWriter fw = new FileWriter(pluginYml)) {
                        fw.write(yamlBuilder.build().toString());
                    } catch (IOException exception) {
                        exception.printStackTrace();
                    }
                }
            } catch (IOException ioException) {
                ioException.printStackTrace();
            }
        });
    }

    private void setupUTF8(Project target) {
        target.getTasks().getByName("compileJava").doFirst(task -> {
            val javaCompile = (JavaCompile) task;
            javaCompile.getOptions().setEncoding("UTF-8");
        });
    }

    private void setupRuntimeDownload(Project target, IcicleAddonData addonData) {
        val runtime = target.getConfigurations().create("runtimeDownload", files -> files.withDependencies(dependencies -> {
            for (Dependency dep : dependencies) {
                val dependency = (ExternalModuleDependency) dep;

                val dependencyGroup = dependency.getGroup();
                val dependencyVersion = dependency.getVersion();
                val dependencyName = dependency.getName();

                if (dependencyVersion == null)
                    addonData.getRuntimeDownloads().add(dependencyGroup + ":" + dependencyName);
                else
                    addonData.getRuntimeDownloads().add(dependencyGroup + ":" + dependencyName + ":" + dependencyVersion);
            }
        }));

        target.getConfigurations().getByName("implementation").extendsFrom(runtime);
    }

    private void setupShadow(Project target, Configuration icicle) {
        val icicleShadow = target.getConfigurations().create("icicleShadow");
        val shadow = target.getConfigurations().create("shadow");

        target.getTasks().getByName("jar").doFirst(task -> {
            val jar = (Jar) task;
            val files = new ArrayList<>(20);

            for (File file : shadow)
                if (file.isDirectory()) files.add(file);
                else files.add(target.zipTree(file));

            for (File file : icicleShadow)
                if (file.isDirectory()) files.add(file);
                else files.add(target.zipTree(file));

            jar.setDuplicatesStrategy(DuplicatesStrategy.INCLUDE);
            jar.from(files.toArray());
        });

        icicle.extendsFrom(icicleShadow);
        target.getConfigurations().getByName("implementation").extendsFrom(shadow);
    }

    private Configuration setupDependencies(Project target, IcicleAddonData addonData) {
        val implementation = target.getConfigurations().getByName("implementation");
        val icicle = target.getConfigurations().create("icicle", files -> files.withDependencies(dependencies -> {
            val newDependencies = new ArrayList<Dependency>(5);
            for (Dependency dep : dependencies) {
                val dependency = (ExternalModuleDependency) dep;

                val dependencyGroup = dependency.getGroup();
                val dependencyVersion = dependency.getVersion();
                if ("icicle".equals(dependencyGroup)) {
                    val dependencyName = dependency.getName();
                    if (dependencyVersion == null) {
                        addonData.getDependencies().add(dependencyName);
                        newDependencies.add(target.getDependencies().create("net.iceyleagons:icicle-addon-" + dependencyName));
                    } else {
                        addonData.getDependencies().add(dependencyName + ":" + dependencyVersion);
                        newDependencies.add(target.getDependencies().create("net.iceyleagons:icicle-addon-" + dependencyName + ":" + dependencyVersion));
                    }
                } else {
                    if (dependencyVersion == null)
                        addonData.getDependencies().add(dependency.getName());
                    else addonData.getDependencies().add(dependency.getName() + ":" + dependencyVersion);
                    newDependencies.add(dependency);
                }
            }

            dependencies.clear();
            dependencies.addAll(newDependencies);
        }));

        implementation.extendsFrom(icicle);
        return icicle;
    }

    private void setupRepository(Project target) {
        IcicleRepositoryHelper.setInstance(new IcicleRepositoryHelper.IcicleRepository(target.getRepositories()));
    }

    private void setupDependencyAliases(Project target) {
        IcicleDependencyHelper.setInstance(new IcicleDependencyHelper.IcicleDependencies(target.getDependencies()));
    }

    private IcicleAddonData setupAddonData(Project target) {
        return (IcicleAddonData) target.getExtensions().create("icicle", IcicleAddonData.class);
    }

    private void setupIcicleGeneration(Project target, IcicleAddonData addonData) {
        val generateIcicleConfig = target.getTasks().register("generateIcicle", IcicleConfigTask.class, task -> {
            task.setGroup("Icicle addon development");
            task.setDescription("Generates the icicle.yml file for the provided addon.");

            // Populate the addonData with required information
            if (addonData.getName() == null)
                addonData.setName(target.getName());
            if (addonData.getVersion() == null)
                addonData.setVersion(target.getVersion().toString());
            if (addonData.getDependencyNotation() == null)
                addonData.setDependencyNotation("net.iceyleagons:icicle-addon-" + addonData.getName().toLowerCase().replace(' ', '-') + ":" + addonData.getVersion());

            task.setData(addonData);
            task.getOutputDirectory().set(target.getLayout().getBuildDirectory().dir(task.getName()));
        });

        target.getTasks().named("processResources", Copy.class, task -> {
            val copyIcicleFile = task.getRootSpec().addChild();
            copyIcicleFile.into("/");
            copyIcicleFile.from(generateIcicleConfig);
        });
    }

    @SneakyThrows
    private YamlMappingBuilder editPluginYml(File file) {
        var yamlBuilder = Yaml.createYamlMappingBuilder();

        val yaml = Yaml.createYamlInput(file).readYamlMapping();
        val existingDepends = yaml.yamlSequence("depend");
        // Copy over existing non-trivial stuff.
        for (YamlNode key : yaml.keys()) {
            if (key.toString().equals("depend")) continue;
            yamlBuilder = yamlBuilder.add(key, yaml.value(key));
        }

        if (existingDepends != null) {
            var sequenceBuilder = Yaml.createYamlSequenceBuilder();
            boolean found = false;
            for (int i = 0; i < existingDepends.size(); ++i) {
                if (!existingDepends.string(i).equalsIgnoreCase("icicle")) continue;
                found = true;
            }

            if (found)
                for (YamlNode yamlNode : existingDepends)
                    sequenceBuilder = sequenceBuilder.add(yamlNode);
            else {
                for (YamlNode yamlNode : existingDepends)
                    sequenceBuilder = sequenceBuilder.add(yamlNode);
                sequenceBuilder = sequenceBuilder.add("Icicle");
            }
            yamlBuilder = yamlBuilder.add("depend", sequenceBuilder.build());
        } else
            yamlBuilder = yamlBuilder.add("depend", Yaml.createYamlSequenceBuilder().add("Icicle").build());

        return yamlBuilder;
    }
}

File: icicle-gradle/src/main/kotlin/net/iceyleagons/gradle/IciclePlugin.kt
package net.iceyleagons.gradle

import net.iceyleagons.gradle.catalogs.IcicleConfiguration
import net.iceyleagons.gradle.catalogs.MinecraftCatalog
import net.iceyleagons.gradle.catalogs.RepositoryCatalog
import net.iceyleagons.gradle.tasks.IcicleConfigTask
import net.iceyleagons.gradle.tasks.PluginYmlTask
import net.iceyleagons.gradle.utils.get
import org.gradle.api.Action
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.Task
import org.gradle.api.artifacts.repositories.UrlArtifactRepository
import org.gradle.api.file.DuplicatesStrategy
import org.gradle.api.tasks.Copy
import org.gradle.api.tasks.compile.JavaCompile
import org.gradle.jvm.tasks.Jar
import java.io.File
import java.net.URL

class IciclePlugin : Plugin<Project> {

    companion object {
        internal var CONF: IcicleConfiguration? = null
    }

    private lateinit var target: Project
    private lateinit var config: IcicleConfiguration

    private fun getDependencyUrls(conf: String): Sequence<Pair<String, URL?>> =
        target.configurations[conf].dependencies.asSequence().mapNotNull {
            it.run { "$group:$name:$version" } to target.repositories.mapNotNull { repo ->
                (repo as? UrlArtifactRepository)?.url
            }.flatMap { repoUrl ->
                "%s/%s/%s/%s/%s-%s".format(
                    repoUrl.toString().trimEnd('/'),
                    it.group?.replace('.', '/') ?: "", it.name, it.version,
                    it.name, it.version
                ).let { x -> listOf("$x.jar", "$x.aar") }
            }.firstNotNullOf { url ->
                kotlin.runCatching {
                    val connection = URL(url).openConnection()
                    connection.getInputStream() ?: throw Exception()
                    connection.url
                    URL(
                        connection.url.toString().replace(
                            "%s/%s/%s/%s-%s".format(
                                it.group?.replace('.', '/') ?: "", it.name, it.version,
                                it.name, it.version
                            ), ""
                        ).removeSuffix(".jar").removeSuffix(".aar")
                    )
                }.getOrNull()
            }
        }

    override fun apply(target_: Project) {
        target = target_
        setup()
        dependencies()
        config()
    }

    private fun config() {
        if (config.modifyPluginYml) {
            val pluginTask = target.tasks.register(
                Strings.TASK_NAME_PLUGIN_YML_MODIFICATION,
                PluginYmlTask::class.java,
                object : Action<PluginYmlTask> {
                    override fun execute(t: PluginYmlTask) {
                        t.description = "Modifies/Creates a plugin.yml in the resources folder of this project."
                        t.group = "Icicle addon development"

                        t.pluginYml = target.file("src/main/resources/plugin.yml")
                    }
                })

            target.tasks["processResources"].dependsOn(pluginTask)
        }

        if (!config.registerIcicleTask)
            return

        val task =
            target.tasks.register(
                Strings.TASK_NAME_ICICLE_YML,
                IcicleConfigTask::class.java,
                object : Action<IcicleConfigTask> {
                    override fun execute(t: IcicleConfigTask) {
                        t.description = "Generates the icicle.yml file for the provided addon."
                        t.group = "Icicle addon development"

                        if (config.registerRuntimeDownload) {
                            for ((dependency, repo) in getDependencyUrls(Strings.CONFIGURATION_RUNTIME_DOWNLOAD))
                                config.dependencies[dependency] = repo.toString().trimEnd('/')

                            for ((dependency, repo) in getDependencyUrls(Strings.CONFIGURATION_ICICLE_ADDON))
                                config.icicle_dependencies[dependency] = repo.toString().trimEnd('/')
                        }

                        CONF = config
                        t.outputDirectory.set(target.layout.buildDirectory.dir(t.name))
                    }
                })

        if (config.generateIcicleYml)
            target.tasks.named("processResources", Copy::class.java, object : Action<Copy> {
                override fun execute(t: Copy) {
                    t.rootSpec.addChild().let {
                        it.duplicatesStrategy = DuplicatesStrategy.EXCLUDE // Hopefully fixes the duplicate files issue.
                        it.into("/")
                        it.from(task)
                    }
                }

            })
    }

    private fun dependencies() {
        val compileOnly = target.configurations[Strings.CONFIGURATION_EXTENDED_CONFIGURATION]

        if (config.registerShadow) {
            val shadow = target.configurations.create(Strings.CONFIGURATION_SHADOW)

            target.tasks["jar"].doFirst(object : Action<Task> {
                override fun execute(t: Task) {
                    t as Jar
                    val files: MutableList<File> = ArrayList(8)

                    for (file in shadow)
                        if (file.isDirectory) files += file
                        else files += target.zipTree(file)

                    t.duplicatesStrategy = DuplicatesStrategy.INCLUDE
                    t.from(files.toTypedArray())
                }

            })
        }

        if (config.registerRuntimeDownload) {
            val runtimeDownload = target.configurations.create(Strings.CONFIGURATION_RUNTIME_DOWNLOAD)
            compileOnly.extendsFrom(runtimeDownload)

            val icicleAddon = target.configurations.create(Strings.CONFIGURATION_ICICLE_ADDON)
            compileOnly.extendsFrom(icicleAddon)
        }
    }

    private fun setup() {
        config = target.extensions.create("icicle", IcicleConfiguration::class.java)

        if (config.registerRepositories)
            target.extensions.create("repos", RepositoryCatalog::class.java)

        if (config.minecraftDependencies)
            target.extensions.create("minecraft", MinecraftCatalog::class.java)

        if (config.forceUtf8)
            target.tasks["compileJava"].doFirst(object : Action<Task> {
                override fun execute(t: Task) {
                    t as JavaCompile
                    t.options.encoding = "UTF-8"
                }

            })
    }


}